src/sss_client/pam_sss_gss.c

```diff
@@ -0,0 +1,477 @@
+    if (ret != EOK) {
+        goto done;
+    }
+
+    *_domain = domain;
+    *_target = target;
+
+done:
+    if (ret != EOK) {
+        free(domain);
+        free(target);
+    }
+
+    return ret;
+}
+
+static errno_t sssd_gssapi_init(pam_handle_t *pamh,
+                                const char *pam_service,
+                                const char *pam_user,
+                                char **_domain,
+                                char **_target)
+{
+    size_t reply_len;
+    uint8_t *reply;
+    errno_t ret;
+
+    ret = sssd_gssapi_init_send(pamh, pam_service, pam_user, &reply,
+                                &reply_len);
+    if (ret != EOK) {
+        return ret;
+    }
+
+    ret = sssd_gssapi_init_recv(pamh, reply, reply_len, _domain, _target);
+    free(reply);
+
+    return ret;
+}
+
+static errno_t sssd_establish_sec_ctx_send(pam_handle_t *pamh,
+                                           const char *pam_service,
+                                           const char *domain,
+                                           const void *gss_data,
+                                           size_t gss_data_len,
+                                           void **_reply,
+                                           size_t *_reply_len)
+{
+    struct sss_cli_req_data req_data;
+    size_t service_len;
+    size_t domain_len;
+    uint8_t *data;
+    int ret_errno;
+    int ret;
+
+    service_len = strlen(pam_service) + 1;
+    domain_len = strlen(domain) + 1;
+
+    req_data.len = (service_len + domain_len) * sizeof(char) + gss_data_len;
+    data = (uint8_t*)malloc(req_data.len);
+    if (data == NULL) {
+        return ENOMEM;
+    }
+
+    memcpy(data, pam_service, service_len);
+    memcpy(data + service_len, domain, domain_len);
+    memcpy(data + service_len + domain_len, gss_data, gss_data_len);
+
+    req_data.data = data;
+    ret = sss_pam_make_request(SSS_GSSAPI_SEC_CTX, &req_data, (uint8_t**)_reply,
+                               _reply_len, &ret_errno);
+    free(data);
+    if (ret != PAM_SUCCESS) {
+        ERROR(pamh, "Communication error [%d, %d]: %s", ret, ret_errno,
+              pam_strerror(pamh, ret));
+        return ret_errno != EOK ? ret_errno : EIO;
+    }
+
+    return ret_errno;
+}
+
+static int sssd_establish_sec_ctx(pam_handle_t *pamh,
+                                  const char *pam_service,
+                                  const char *domain,
+                                  const char *target)
+{
+    bool established;
+    gss_ctx_id_t ctx = GSS_C_NO_CONTEXT;
+    gss_buffer_desc input = GSS_C_EMPTY_BUFFER;
+    gss_buffer_desc output = GSS_C_EMPTY_BUFFER;
+    OM_uint32 flags = GSS_C_MUTUAL_FLAG;
+    gss_name_t gss_name;
+    OM_uint32 ret_flags;
+    OM_uint32 major;
+    OM_uint32 minor;
+    int ret;
+
+    ret = string_to_gss_name(pamh, target, &gss_name);
+    if (ret != 0) {
+        return ret;
+    }
+
+    /* Do the handshake. */
+    established = false;
```

Comment:
Hoist.

Reply:
I'm sorry but I don't understand what's this comment about.

Reply:
This variable is no longer present with the switch to `do while`